### PR TITLE
Improved infrastructure for `FromRecord` and `Mutator` traits.

### PR DESCRIPTION
**Pull request by ryzhyk**
See commit messages for details.

### CODE REVIEW

**Review by Kixiron — APPROVED** [review]
LGTM, maybe add some tests to `rust/template/ddlog_derive/tests/ui`?

**Comment by ryzhyk** [comment]
> LGTM, maybe add some tests to `rust/template/ddlog_derive/tests/ui`?

Will do, thanks.

File: test/datalog_tests/map_test.dat [hunk 1]
```diff
@@ -2,3 +2,29 @@ dump map_test::MapTransforms;
...
+modify map_test::NestedMap 1 <- map_test::NestedMap{1, [
+    (0, [("0", "0")]), 
+    (1, [("1", "b"), ("2", ()), ("30", "30")])
+]};
+
+commit dump_changes;
+
+start;
+
+modify map_test::NestedMap 1 <- map_test::NestedMap{1, [
```
Comment (mbudiu-vmw) [inline test/datalog_tests/map_test.dat:26]: shouldn't this be documented in the tutorial?

File: rust/template/ddlog_derive/tests/ui/pass/mutator.rs [hunk 2]
```diff
@@ -32,3 +36,14 @@ impl Default for Enum {
...
+#[derive(FromRecord, IntoRecord, Mutator, Debug, Clone, PartialEq, Deserialize)]
+struct NestedMaps {
+    m: BTreeMap<BTreeMap<u64, String>, BTreeMap<String, BTreeMap<u32, u32>>>
+}
+
+#[derive(FromRecord, IntoRecord, Mutator, Debug, Clone, PartialEq, Deserialize)]
+enum ContainersEnum {
+    Map(BTreeMap<Enum, Struct>),
+    Vec(Vec<NestedMaps>)
+}
```
Comment (Kixiron) [inline rust/template/ddlog_derive/tests/ui/pass/mutator.rs:49]: You can run actual tests within the compile tests by writing code in the `main()` function, I'd move this to another file and write some expected behavior assertions in the main function

Comment (ryzhyk) [reply rust/template/ddlog_derive/tests/ui/pass/mutator.rs]: Done, thanks!

**Comment by ryzhyk** [comment]
@mbudiu-vmw , does this help with #1038? If so, I'll merge once the pipeline succeedds.

**Pull request merged**